pandas/core/arrays/sparse/scipy_sparse.py

```diff
@@ -19,6 +19,20 @@ def _check_is_partition(parts, whole):
         raise ValueError("Is not a partition because union is not the whole.")
 
 
+def _levels_to_axis(ss, levels, valid_ilocs, sort_labels=False):
```

Comment:
can you type & add a doc-string

Reply:
we agree I should revert e09d760 right?

Reply:
I'd shown the perf in  https://github.com/pandas-dev/pandas/pull/42925#discussion_r684752285, which was worse that what I'd initially obtained. I just looked into it and turns out most of the time was lost on `_ensure_arraylike` because of the input I fed `factorize`. I dug a little and from the `_values` attribute of `MultiIndex`, I found that `lib.fast_zip` would enable me to build an input that's convenient for `factorize` (see `621b102`). This even sped things up a little compared to the first solution I proposed.